File: examples/basic_memory.py
"""
Example: Basic Memory CRUD

Demonstrates saving, listing, searching, and deleting memories
using the MemoryOS Python SDK.

Run: python3 examples/basic_memory.py
"""

import sys
import os

# Resolve the local memoryos package (no pip install needed)
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "packages", "memoryos-py"))

from memoryos import MemoryOS


def main():
    # 1. Connect to MemoryOS
    client = MemoryOS(
        api_url="http://localhost:3000",
        agent_id="agent_example_bot",
    )
    print(f"Connected: {client}\n")

    # 2. Register the agent (first time only)
    try:
        identity = client.identity.register("Example Bot")
        print(f"✓ Agent registered: {identity.get('agent', {}).get('name')}")
    except Exception as e:
        print(f"  Agent already exists or error: {e}")

    # 3. Save memories of different types
    print("\n── Saving Memories ──")

    client.memory.save(
        "User prefers dark mode and compact layouts",
        type="semantic",
        tags=["preference", "ui"],
        importance=4,
    )
    print("  ✓ Saved semantic memory (preference)")

    client.memory.save(
        "On May 1, deployed v2.0 to production with zero downtime",
        type="episodic",
        tags=["deployment", "milestone"],
        importance=5,
    )
    print("  ✓ Saved episodic memory (event)")

    client.memory.save(
        "Deploy checklist: 1) Run tests 2) Build 3) Stage 4) Smoke test 5) Promote",
        type="procedural",
        tags=["deployment", "process"],
        importance=4,
    )
    print("  ✓ Saved procedural memory (workflow)")

    # 4. List all memories
    print("\n── Listing Memories ──")
    result = client.memory.list(limit=10)
    memories = result.get("memories", [])
    print(f"  Found {len(memories)} memories:")
    for m in memories:
        print(f"    [{m['type']}] {m['content'][:60]}...")

    # 5. Semantic search
    print("\n── Semantic Search ──")
    search = client.memory.search("how to deploy")
    hits = search.get("memories", [])
    print(f"  Query: 'how to deploy' → {len(hits)} results")
    for hit in hits[:3]:
        print(f"    [{hit['type']}] {hit['content'][:60]}...")

    # 6. Platform status
    print("\n── Platform Status ──")
    status = client.status()
    print(f"  0G Configured: {status.get('configured')}")
    print(f"  Total Memories: {status.get('totalMemories', '?')}")

    client.close()
    print("\n✅ Done!")


if __name__ == "__main__":
    main()

File: examples/basic_memory.mjs
/**
 * Example: Basic Memory CRUD with the JS SDK
 *
 * Demonstrates saving, listing, searching, and using skills
 * with the MemoryOS OpenClaw JS SDK.
 *
 * Run: node examples/basic_memory.mjs
 */

import { MemoryOSClient } from "memoryos-openclaw";

const client = new MemoryOSClient({
  apiUrl: "http://localhost:3000",
  agentId: "agent_example_js",
});

console.log("── MemoryOS JS SDK Example ──\n");

// 1. Register agent
try {
  const identity = await client.identity.register("JS Example Agent");
  console.log(`✓ Agent registered: ${identity.agent?.name || client.agentId}`);
} catch (e) {
  console.log(`  Agent exists or error: ${e.message}`);
}

// 2. Save memories
console.log("\n── Saving Memories ──");

await client.memory.save("User prefers TypeScript over JavaScript", {
  type: "semantic",
  tags: ["preference", "language"],
  importance: 4,
});
console.log("  ✓ Saved semantic memory");

await client.memory.save("Migrated database from Postgres to Supabase on April 28", {
  type: "episodic",
  tags: ["migration", "database"],
  importance: 5,
});
console.log("  ✓ Saved episodic memory");

// 3. List memories
console.log("\n── Listing Memories ──");
const list = await client.memory.list({ limit: 5 });
console.log(`  Found ${list.count} memories:`);
for (const m of list.memories) {
  console.log(`    [${m.type}] ${m.content.slice(0, 60)}...`);
}

// 4. Semantic search
console.log("\n── Semantic Search ──");
const search = await client.memory.search("database changes");
console.log(`  Query: 'database changes' → ${search.count} results`);
for (const m of search.memories.slice(0, 3)) {
  console.log(`    [${m.type}] ${m.content.slice(0, 60)}...`);
}

// 5. List skills
console.log("\n── Skills Marketplace ──");
const skills = await client.skills.list();
console.log(`  Found ${skills.skills.length} skills:`);
for (const s of skills.skills.slice(0, 3)) {
  console.log(`    [${s.category}] ${s.name} — ${s.price} OG`);
}

// 6. Platform status
console.log("\n── Status ──");
const status = await client.status();
console.log(`  0G Configured: ${status.configured}`);

client.close();
console.log("\n✅ Done!");
